Task/add quartz job scheduling

Jamie-Unosquare opened the pull request:
Scheduling for 3am runs and prune job event implemented.

Would like advice on the best way to implement a way for the system to know which job is tied to which trigger, should there be a DB for jobs that is linked to a trigger, or an enum for jobs linked up to a case statement?

Jamie-Unosquare committed:
Initial Job Scheduling Commit

Implemented Scheduling Framework
Implemented DB Schema / Relations
Implemented Job Triggering
Added dependency injection
Refactored old DI to use optional parameters

Todo: Implement Job functionality

Jamie-Unosquare committed:
Removed unused file

Jamie-Unosquare committed:
Added main method for Scheduled Jobs
Added ReadKey for console monitoring

Jamie-Unosquare committed:
Removed unused variable

Jamie-Unosquare committed:
Added scheduled job

Jamie-Unosquare committed:
Added job to prune events

Jamie-Unosquare committed:
Merge branch 'Task/Add-Quartz-Job-Scheduling' of https://github.com/UnosquareBelfast/admin-backend into Task/Add-Quartz-Job-Scheduling

Jamie-Higgins commented on AdminCore.ScheduledJobs/SchedulingManager.cs:
```diff
@@ -0,0 +1,10 @@
+  {
+    private static void Main()
+    {
+      //TODO remove unused main method
```
What is the best way to handle this? Unused main method but project needs it to compile

Jamie-Unosquare committed:
Removed unused main method class

Jamie-Unosquare committed:
Removed unecessary brakcets in DB populate script

BrownMark commented on AdminCore.DatabaseMigration/Program.cs:
```diff
@@ -12,7 +12,7 @@ public static class Program
     public static void Main(string[] args)
     {
-      DependencyInjection.RegisterDependencyInjection();
+      DependencyInjection.RegisterDependencyInjection(null, null);
```
Can we overload this? Seems to be a lot of bulls passed into this method with usage

BrownMark commented on AdminCore.ScheduledJobs/ScheduleRunner.cs:
```diff
@@ -0,0 +1,73 @@
+
+    private static IJobDetail GetJob(ScheduleDto schedule)
+    {
+      return schedule.IsActive ? GetJobForScheduler<StartupJob>(schedule) : GetJobForScheduler<ShutdownJob>(schedule);
```
Don’t need shutdown and can we rename startup to PruneExpiredEventsJob.

Can we also have a jobs enum and switch on it? Just cleaner to maintain.

BrownMark commented on AdminCore.ScheduledJobs/ShutdownJob.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using Quartz;
+using System.Threading.Tasks;
+
+namespace AdminCore.ScheduledJobs
```
Pew pew

BrownMark commented on AdminCore.ScheduledJobs/ScheduleRunner.cs:
```diff
@@ -0,0 +1,73 @@
+      _scheduler = factory.GetScheduler().Result;
+      _scheduler.Start();
+
+      _scheduler.JobFactory = new DatabaseJobFactory(databaseContext);
```
So this is a singleton and the same DB context will be used for each job. This was fine when the last approach used LikeDB. Here we might get unit of work issues

BrownMark reviewed (changes requested):
We probably need to get back to unit testing too

Jamie-Unosquare committed:
Changes from PR

Overloaded DI
New DB scope
Renamed Startup Job
Removed Shutdown

BrownMark commented on AdminCore.ScheduledJobs/ScheduleRunner.cs:
```diff
@@ -38,7 +38,10 @@ public void StartSchedule(ScheduleDto schedule)
     private static IJobDetail GetJob(ScheduleDto schedule)
     {
-      return schedule.IsActive ? GetJobForScheduler<StartupJob>(schedule) : GetJobForScheduler<ShutdownJob>(schedule);
+      if (schedule.IsActive)
```
please add {}

BrownMark commented on AdminCore.Services/Configuration/DependencyInjection.cs:
```diff
@@ -18,8 +18,7 @@ public static class DependencyInjection
 
-    public static void RegisterDependencyInjection(IServiceCollection serviceDescriptor = null,
-      params ServiceDescriptor[] serviceDescriptors)
+    public static void RegisterDependencyInjection(IServiceCollection serviceDescriptor = null)
```
the idea here was to have two signatures call mostly the same implementation, rather than the parameters dictating which mappings are set. I'll grab some time with you and go over it.

BrownMark commented on AdminCore.ScheduledJobs/PruneExpiredEventsJob.cs:
```diff
@@ -0,0 +1,73 @@
+  {
+    public Task Execute(IJobExecutionContext context)
+    {
+      var schedules = DatabaseContext.SchedulesRepository.Get(x => x.IsActive);
```
This returns an empty collection if nothing is found. You'll need to check the count.

BrownMark commented on AdminCore.ScheduledJobs/PruneExpiredEventsJob.cs:
```diff
@@ -0,0 +1,73 @@
+      return unapprovedEvent;
+    }
+
+    private List<EventDate> GetExpiredUnapprovedEvents(int awaitingApproval)
```
for a method like get Unexpired unapproved events you shouldn't be passing the type in. It should be within the method.

BrownMark commented on AdminCore.ScheduledJobs/PruneExpiredEventsJob.cs:
```diff
@@ -0,0 +1,73 @@
+
+    private Event CancelUnapprovedEvent(EventDate expiredUnapprovedEvent)
+    {
+      var cancelledEventStats = (int)EventStatuses.Cancelled;
```
cancelledEventStatus

BrownMark commented on AdminCore.ScheduledJobs/PruneExpiredEventsJob.cs:
```diff
@@ -0,0 +1,73 @@
+      if (expiredUnapprovedEvents.Any())
+      {
+        SetExpiredUnapprovedEventsStatusToCancelled(expiredUnapprovedEvents);
+        Console.WriteLine("Successfully pruned " + expiredUnapprovedEvents.Count +
```
string interpolation is your friend.

BrownMark commented on AdminCore.JobExecuter/Program.cs:
```diff
@@ -0,0 +1,25 @@
+    private static void Main(string[] args)
+    {
+      DependencyInjection.RegisterDependencyInjection(
+        new ServiceDescriptor(typeof(IScheduleRunner), typeof(ScheduleRunner), ServiceLifetime.Singleton)
```
this is still going to cause unit of work issues.

Jamie-Unosquare commented:
Would it be possible to get some time to go over the best way to mock the database? This seems to have been the limiting factor before when we were doing testing and the MockDatabase that Niall made is / will be outdated with the changes we implement,

Jamie-Unosquare committed:
Pull Request Changes:

Overloaded RegisterDependencyInjection
Check count on Schedules
Added String interpolation
Added curly brackets to IF statement
AdminCore Dependency Injection is Scoped
Job Scheduling Dependency Injection is Transient.

Jamie-Unosquare commented on AdminCore.ScheduledJobs/DatabaseJobFactory.cs:
```diff
@@ -0,0 +1,32 @@
+      var job = (IJob)Activator.CreateInstance(bundle.JobDetail.JobType);
+      if (job is BaseDatabaseJob)
+      {
+        ((BaseDatabaseJob)job).DatabaseContext = _databaseContext;
```
use service locator for db context

Jamie-Unosquare commented on AdminCore.DatabaseMigration/db/migrations/V1.10__add_schedule_jobs.sql:
```diff
@@ -0,0 +1,25 @@
+/*
+                                   SCHEDULE JOB TABLE
+*/
+  ----------------------------------------------------------------------------------------
```
lines

Jamie-Unosquare commented on AdminCore.Services/Configuration/DependencyInjection.cs:
```diff
@@ -44,22 +45,48 @@ public static void RegisterDependencyInjection(IServiceCollection services = nul
-    private readonly IServiceProvider _container;
 
-    internal DependencyInjectionContainer(IServiceProvider container)
+    public static void RegisterDependencyInjection(params ServiceDescriptor[] serviceDescriptors)
```
rename

Jamie-Unosquare committed:
Changed Overload to Seperate DI
Added ServiceLocator DB instance
Fromatted SQL Lines
Re-implemented paramaterless constructor

BrownMark reviewed (approved):
Good work man

Jamie-Unosquare committed:
Merge branch 'develop' into Task/Add-Quartz-Job-Scheduling
